cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -98,11 +131,85 @@ def start_master(etcd):
-    hookenv.status_set('active', 'Kubernetes master running.')
+    hookenv.status_set('active', 'Kubernetes master services ready.')
+
+
+@when('kube-dns.available', 'cluster-dns.connected', 'sdn-plugin.available')
+def send_cluster_dns_detail(cluster_dns, sdn_plugin):
+    details = sdn_plugin.get_sdn_config()
+    sdn_ip = get_dns_ip(details['cidr'])
+    cluster_dns.set_dns_info(53, hookenv.config('dns_domain'), sdn_ip)
+
+
+# TODO: This needs a much better relationship name...
+@when('kube-api-endpoint.available')
+def push_service_data(kube_api):
+    ''' Send configuration to the load balancer, and close access to the
+    public interface '''
+    hookenv.close_port(8080)
+    kube_api.configure(port=8080)
+
+
+@when('kube-api.connected')
+def push_api_data(kube_api):
+    ''' Send configuration to remote consumer'''
+    data = {'private_address': hookenv.unit_private_ip()}
+
+    # TODO Replace with actual TLS interface code
+    if not is_state('ca.connected'):
+        data['tls'] = False
+        data['port'] = 8080
+
+    kube_api.set_api_credentials(data['private_address'], data['port'],
+                                 data['tls'])
+
+
+@when('kube_master_components.installed', 'sdn-plugin.available')
+def gather_sdn_data(sdn_plugin):
+    sdn_data = sdn_plugin.get_sdn_config()
+    if not sdn_data['cidr'] or not sdn_data['subnet'] or not sdn_data['mtu']:
+        hookenv.status_set('waiting', 'Waiting on SDN configuration')
+        return
+    api_opts = FlagManager('kube-apiserver')
+    api_opts.add('--service-cluster-ip-range', sdn_data['cidr'])
+    set_state('kube-sdn.configured')
+
+
+@when('kube_master_components.installed')
+@when_not('kubernetes.dashboard.available')
+def launch_kubernetes_dashboard():
+    ''' Launch the Kubernetes dashboard. If not enabled, attempt deletion '''
+    if hookenv.config('dashboard'):
+        # TODO - make this self contained
+        dashboard_manifest = 'https://rawgit.com/kubernetes/dashboard/master/src/deploy/kubernetes-dashboard.yaml'  # noqa
+        cmd = ['kubectl', 'create', '-f', dashboard_manifest]
+        call(cmd)
+        set_state('kubernetes.dashboard.available')
+    else:
+        cmd = ['kubectl', 'delete', 'kubernetes-dashboard',
+               '--namespace=kube-system']
+        try:
+            call(cmd)
+        except CalledProcessError:
+            pass
+
+
+@when('kube_master_components.installed', 'kube-sdn.configured',
+      'sdn-plugin.available')
+@when_not('kube-dns.started')
+def start_kube_dns(sdn_plugin):
+    ''' State guard to starting DNS '''
+    context = prepare_sdn_context(sdn_plugin)
+    render('kubedns-rc.yaml', '/etc/kubernetes/addons/kubedns-rc.yaml',
+           context)
+    render('kubedns-svc.yaml', '/etc/kubernetes/addons/kubedns-svc.yaml',
+           context)
+    # This should be auto-loaded by the addon manager, but it doesnt appear
+    # to do so.
+    launch_dns()
+    set_state('kube-dns.available')
```

Comment:
also the launch_dns() method uses "kubedns.available" not "kube-dns.available" we should fix them all.